File: src/weeb.mjs
import * as Message from "../lib/Message.mjs";
import * as Random from "../lib/Random.mjs";
import * as Dataset from "../data/index.mjs";

import Discord from "discord.js";
import snekfetch from "snekfetch";

import util from "util";

export const boobpat = Message.typing(async (message, content) => await message.channel.send(new Discord.RichEmbed({
	description: `${message.author} patted ${content ? `${content} on the` : "some"} boobs!`,
	image: { url: "https://cdn.discordapp.com/attachments/403299886352695297/472762872209080321/image.gif" },
})));

export const buttpat = Message.typing(async (message, content) => await message.channel.send(new Discord.RichEmbed({
	description: `${message.author} patted ${content || "Yuzu"} on the butts!`,
	image: { url: "https://78.media.tumblr.com/165f23ece178a17968de50f084a9ecec/tumblr_p25cyprR041vptudso2_400.gif" },
})));

export const feed = Message.typing(async (message, content) => await message.channel.send(new Discord.RichEmbed({
	description: `${message.author} fed ${content || "a random anime character"}!`,
	image: (await snekfetch.get("https://nekos.life/api/v2/img/feed")).body,
})));

export const hug = (message, content) => message.channel.send(new Discord.RichEmbed({
	description: `${message.author} hugged ${content || "Yuri"}!`,
	image: { url: Random.pick(Dataset.hugs) },
}));

export const kiss = (message, content) => message.channel.send(new Discord.RichEmbed({
	description: `${message.author} kissed ${content || "Natsuki"}!`,
	image: { url: Random.pick(Dataset.kisses) },
}));

export const lewd = message => message.channel.send("https://youtu.be/qr89xoZyE1g");

export const lick = (message, content) => message.channel.send(new Discord.RichEmbed({
	description: `${message.author} licked ${content || "the air"}!`,
	image: { url: "https://cdn.discordapp.com/attachments/421196261132075009/421920949277818891/LickTemplate.gif" },
}));

export const licc = lick;

export const neko = Message.typing(async (message, content) =>
{
	const endpoint = message.channel.nsfw ? "https://nekos.life/api/v2/img/lewd" : "https://nekos.life/api/v2/img/neko";

	return await message.channel.send(new Discord.RichEmbed({
		description: "Here comes your random neko.",
		image: (await snekfetch.get(endpoint)).body,
	}));
});

export const smash = (message, content) => message.channel.send(new Discord.RichEmbed({
	description: `${message.author} smashed${content && " "}${content}!`,
	image: { url: "https://raw.githubusercontent.com/jdh8/natsuki/master/assets/smash.png" },
}));

File: src/weeb.rs
use crate::Context;
use anyhow::Context as _;
use poise::serenity_prelude as serenity;

/// Fetch an image URL from an API that responds with `{"url": "..."}`
/// (e.g. nekos.life, otakugifs.xyz) and send it as an embed.
async fn send_image_embed(
    ctx: Context<'_>,
    endpoint: &str,
    description: String,
) -> anyhow::Result<()> {
    let json = ctx
        .data()
        .http
        .get(endpoint)
        .send()
        .await?
        .json::<serde_json::Value>()
        .await?;
    let url = json["url"].as_str().context("Invalid image URL")?;

    ctx.send(poise::CreateReply {
        embeds: vec![serenity::CreateEmbed::new()
            .description(description)
            .image(url)],
        ..Default::default()
    })
    .await?;
    Ok(())
}

/// Feed someone
///
/// Feed someone or a random anime character
///
/// **Usage:** /feed [user|text]
#[poise::command(category = "Weeb", slash_command)]
pub async fn feed(
    ctx: Context<'_>,
    #[description = "Someone to feed"] text: Option<String>,
) -> anyhow::Result<()> {
    let text = text.as_deref().unwrap_or("a random anime character");
    let description = ctx.author().to_string() + " fed " + text + "!";
    send_image_embed(ctx, "https://nekos.life/api/v2/img/feed", description).await
}

/// Hug someone
///
/// Hug someone or Yuri
///
/// **Usage:** /hug [user|text]
#[poise::command(category = "Weeb", slash_command)]
pub async fn hug(
    ctx: Context<'_>,
    #[description = "Someone to hug"] text: Option<String>,
) -> anyhow::Result<()> {
    let text = text.as_deref().unwrap_or("a random anime character");
    let description = ctx.author().to_string() + " hugged " + text + "!";
    send_image_embed(
        ctx,
        "https://api.otakugifs.xyz/gif?reaction=hug",
        description,
    )
    .await
}

/// Kiss someone
///
/// Kiss someone or Natsuki
///
/// **Usage:** /kiss [user|text]
#[poise::command(category = "Weeb", slash_command)]
pub async fn kiss(
    ctx: Context<'_>,
    #[description = "Someone to kiss"] text: Option<String>,
) -> anyhow::Result<()> {
    let text = text.as_deref().unwrap_or("a random anime character");
    let description = ctx.author().to_string() + " kissed " + text + "!";
    send_image_embed(
        ctx,
        "https://api.otakugifs.xyz/gif?reaction=kiss",
        description,
    )
    .await
}

/// Display a video
///
/// Display a lewd video
///
/// **Usage:** /lewd
#[poise::command(category = "Weeb", slash_command)]
pub async fn lewd(ctx: Context<'_>) -> anyhow::Result<()> {
    ctx.say("https://youtu.be/qr89xoZyE1g").await?;
    Ok(())
}

/// Lick someone
///
/// Lick someone or the air
///
/// **Usage:** /lick [user|text]
#[poise::command(category = "Weeb", slash_command)]
pub async fn lick(
    ctx: Context<'_>,
    #[description = "Someone to lick"] text: Option<String>,
) -> anyhow::Result<()> {
    let text = text.as_deref().unwrap_or("the air");
    let description = ctx.author().to_string() + " licked " + text + "!";
    send_image_embed(
        ctx,
        "https://api.otakugifs.xyz/gif?reaction=lick",
        description,
    )
    .await
}

/// Show a random neko
///
/// Show a random neko from nekos.life
///
/// **Usage:** /neko
#[poise::command(category = "Weeb", slash_command)]
pub async fn neko(ctx: Context<'_>) -> anyhow::Result<()> {
    send_image_embed(
        ctx,
        "https://nekos.life/api/v2/img/neko",
        "Here comes your random neko.".to_owned(),
    )
    .await
}
